File: src/main/scala/edu/agh/lroza/actors/scala/LoginActor.scala
package edu.agh.lroza.actors.scala

import java.util.UUID
import edu.agh.lroza.scalacommon.Problem
import akka.actor.{ActorRef, Actor}
import edu.agh.lroza.actors.scala.Classes._

class LoginActor(noticesActor: ActorRef) extends Actor {
  var loggedUsers = Set[UUID]()

  def login(username: String, password: String): Either[Problem, UUID] = if (username == password) {
    val token = UUID.randomUUID()
    loggedUsers = loggedUsers + token
    Right(token)
  } else {
    Left(Problem("Wrong password"))
  }

  def logout(token: UUID): Option[Problem] = if (loggedUsers.contains(token)) {
    loggedUsers = loggedUsers - token
    None
  } else {
    Some(Problem("Invalid token"))
  }

  val notLogged = Left(Problem("Invalid token. Please log in."))

  def validateToken(token: UUID)(code: => Unit) {
    if (loggedUsers.contains(token)) {
      code
    } else {
      self reply notLogged
    }
  }

  protected def receive = {
    case message@NoticesMessage(token) => validateToken(token) {
      noticesActor.!(message)(self.getChannel)
    }
    case message@NoticeMessage(token, ActorId(noticeActor)) => validateToken(token) {
      if (!noticeActor.tryTell(message)(self.getChannel)) {
        self reply leftDeletedNotice
      }
    }
    case NoticeMessage(token, id) => validateToken(token) {
      self reply Left(Problem("There is no such notice '" + id + "'"))
    }
    case Login(username, password) => self reply login(username, password)
    case Logout(token) => self reply logout(token)
  }
}

File: src/main/java/edu/agh/lroza/actors/java/LoginActor.java
package edu.agh.lroza.actors.java;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

import akka.actor.ActorRef;
import akka.actor.Channel;
import akka.actor.UntypedActor;
import edu.agh.lroza.actors.java.NoticeActor.NoticeActorMessage;
import edu.agh.lroza.actors.java.NoticesActor.NoticesActorMessage;
import edu.agh.lroza.javacommon.ProblemException;

@SuppressWarnings({"unchecked", "rawtypes"})
public class LoginActor extends UntypedActor {
    private Set<UUID> loggedUsers = new HashSet<>();
    private Channel noticesActor;

    public LoginActor(ActorRef noticesActor) {
        this.noticesActor = noticesActor;
    }

    public static class Login {
        private final String username;
        private final String password;

        public Login(String username, String password) {
            this.username = username;
            this.password = password;
        }
    }

    public static class Logout {
        private final UUID token;

        public Logout(UUID token) {
            this.token = token;
        }
    }

    public void onReceive(Object message) {
        if (message instanceof NoticeActorMessage) {
            NoticeActorMessage noticeMessage = (NoticeActorMessage) message;
            if (loggedUsers.contains(noticeMessage.getToken())) {
                Channel noticeActor = noticeMessage.getActor();
                if (noticeActor == null || !noticeActor.tryTell(noticeMessage, getContext().getChannel())) {
                    getContext().reply(new ProblemException("There is no such notice"));
                }
            } else {
                getContext().reply(new ProblemException("Please log in"));
            }
        } else if (message instanceof NoticesActorMessage) {
            NoticesActorMessage noticeMessage = (NoticesActorMessage) message;
            if (loggedUsers.contains(noticeMessage.getToken())) {
                noticesActor.tell(noticeMessage, getContext().getChannel());
            } else {
                getContext().reply(new ProblemException("Please log in"));
            }
        } else if (message instanceof Login) {
            Login login = (Login) message;
            if (login.username.equals(login.password)) {
                UUID token = UUID.randomUUID();
                loggedUsers.add(token);
                getContext().reply(token);
            } else {
                getContext().reply(new ProblemException("Wrong password"));
            }
        } else if (message instanceof Logout) {
            Logout logout = (Logout) message;
            if (loggedUsers.contains(logout.token)) {
                loggedUsers.remove(logout.token);
                getContext().reply(null);
            } else {
                getContext().reply(new ProblemException("Invalid token"));
            }
        } else {
            throw new IllegalArgumentException("Unknown message: " + message);
        }
    }
}